packages/adapter-components/src/client/rate_limiter.ts

```diff
@@ -191,17 +246,72 @@ export class RateLimiter {
 
   /**
    * Wraps a given task with counters bookkeeping.
-   * @param task The task to wrapped with bookkeeping.
-   * @returns A function that returns a promise resolving to the result of the task
+   * @param task The task to be wrapped with bookkeeping.
+   * @returns A function that returns a promise resolving to the result of the task.
    */
-  private wrapWithBookKeeping<T>(task: () => Promise<T>): () => Promise<T> {
+  private wrapWithBookKeeping<T>(task: TaskType<T>, isRetry: boolean = false): TaskType<T> {
     return async (): Promise<T> => {
       this.internalCounters.pending -= 1
       this.internalCounters.running += 1
-      const res = await task()
-      this.internalCounters.running -= 1
-      this.internalCounters.done += 1
-      return res
+      if (isRetry) {
+        this.internalCounters.retries += 1
+      }
+      let res
+      try {
+        res = await task()
+        this.internalCounters.succeeded += 1
+        return res
+      } catch (e) {
+        this.internalCounters.failed += 1
+        throw e
+      } finally {
+        this.internalCounters.running -= 1
+        this.internalCounters.done += 1
+      }
+    }
+  }
+
+  private wrapWithPauseOnFail<T>(task: TaskType<T>, numAttempts: number): TaskType<T> {
+    return async () => {
+      try {
+        return await task()
+      } catch (e) {
+        if (this.internalOptions.retryPredicate(numAttempts, e)) {
+          const delay = this.internalOptions.calculateRetryDelayMS(numAttempts, e)
+          if (delay > 0) {
+            this.pause(delay)
+          }
+        }
+        throw e
+      }
+    }
+  }
+
+  private async recursiveAdd<T>(task: TaskType<T>, numAttempts: number): TaskReturnType<T> {
+    let res: TaskReturnType<T>
+    this.internalCounters.total += 1
+    this.internalCounters.pending += 1
+    const bookKeepingWrap = this.wrapWithBookKeeping(this.getDelayedTask(task), numAttempts > 1)
```

Comment:
it's more intuitive to start counting the attempts from 0 IMO
```suggestion
    const bookKeepingWrap = this.wrapWithBookKeeping(this.getDelayedTask(task), numAttempts > 0)
```

Reply:
Done